cpp/src/plasma/client.cc

```diff
@@ -313,8 +313,9 @@ bool PlasmaClient::Impl::IsInUse(const ObjectID& object_id) {
 int PlasmaClient::Impl::GetStoreFd(int store_fd) {
   auto entry = mmap_table_.find(store_fd);
   if (entry == mmap_table_.end()) {
-    int fd = recv_fd(store_conn_);
-    ARROW_CHECK(fd >= 0) << "recv not successful";
+    int fd;
+    auto status = store_conn_->RecvFd(&fd);
+    ARROW_CHECK(status.ok() && fd >= 0) << "recv not successful";
```

Comment:
There are about 69 `ARROW_CHECK`s that should be replaced with a status code in plasma, so I think they should be addressed in another PR.

Reply:
Please be mindful about distinguishing different cases:

* Assertions that "cannot fail" (unless the _implementation_ is wrong). We use DCHECK/ARROW_CHECK in many cases to assert invariants to make it easier to refactor and sanity check things in debug builds. For example, if you have `ARROW_CHECK(i >= 0)= and you know that `i` can never be 0 (unless you have a bug in your code) then it would not be appropriate to return error Status in that case
* Failures as the result of truly exceptional behavior, or in limited cases unsanitized user input

Reply:
Agreed that care should be taken here. One case that we have met is the handling for plasma store crash. Because there are a couple of checks where client sends a certain request and expects the same kind of reply. If the store crashes in this case the client would get a DisconnectClient I think and then crash with a core dump. As there are many actors in a given machine, this would result in like 10 cores from workers and 1 from raylet.

In this case would it make sense to just let the process using plasma client exit with an error code, and a log saying that it exits because plasma store dies?

Reply:
I have modified the code according to @wesm 's advice.